test/fuzz/transformation_add_float_type_test.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_add_type_float.h"
+
+#include "gtest/gtest.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "test/fuzz/fuzz_test_util.h"
+
+namespace spvtools {
+namespace fuzz {
+namespace {
+
+TEST(TransformationAddConstantBooleanTest, NeitherPresentInitiallyAddBoth) {
+  std::string shader = R"(
+               OpCapability Shader
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %4 "main"
+               OpExecutionMode %4 OriginUpperLeft
+               OpSource ESSL 310
+               OpName %4 "main"
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+               OpReturn
+               OpFunctionEnd
+  )";
+
+  const auto env = SPV_ENV_UNIVERSAL_1_3;
+  const auto consumer = nullptr;
+  const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+  
+  // Should be able to add a 32-bit float type with id 7
+  ASSERT_TRUE(TransformationAddTypeFloat(7, 32)
+                  .IsApplicable(context.get(), transformation_context));
+  
+
+  // Should be able to add a 16-bit float type with id 7
+  if (!context.get()->get_feature_mgr()->HasCapability(SpvCapabilityFloat16)) {
+       ASSERT_FALSE(TransformationAddTypeFloat(7, 16)
+          .IsApplicable(context.get(), transformation_context));
+  }
```

Comment:
The above SPIR-V does *not* declare the float 16 capability. Thus the if condition will be false, and the assertion will not be checked.

Instead, you should either:

- include the float 16 capability in the SPIR-V and assert that the transformation is applicable
or
- do not include the float 16 capability in the SPIR-V and assert that the transformation is not applicable

It would be good to have separate tests for these - e.g. a test that asserts that you cannot add float type with interesting bit-widths if the capabilities are not there, and another test showing that you can add them if the capabilities are present.

Could you try that?